### PR TITLE
bpo-31356: Fix multiple errors in gc.ensure_disabled()

### PR DESCRIPTION
**Pull request by 1st1**
https://bugs.python.org/issue31356

### CODE REVIEW

**Comment by 1st1** [comment]
cc @lisroach

**Comment by 1st1** [comment]
@ned-deily If there's time, when this one is reviewed by @lisroach and @serhiy-storchaka it would be great if this one goes into beta-1.

Currently, running `$ ./python -We -m test test_gc` SIGABRTs on master.

File: Modules/gcmodule.c
Comment (vstinner) [inline Modules/gcmodule.c]: Don't you have to DECREF somewhere? Try: ./python -m test -R 3:3 test_gc

**Comment by vstinner** [comment]
cc @pablogsal

File: Modules/gcmodule.c
Comment (1st1) [reply Modules/gcmodule.c]: Good catch; fixed.

**Comment by 1st1** [comment]
I don't understand the "test_ensure_disabled_thread" test: why is `thread_inside_status` true?

**Comment by 1st1** [comment]
Why are we using `PyGILState_Release` in the code?

File: Lib/test/test_gc.py
Comment (serhiy-storchaka) [inline Lib/test/test_gc.py:1057]: What line emits a warning? It would be better to surround with assertWarnsRegex() only that line. Or it can be emitted by arbitrary line in the block?

Note that warnings emitted in other thread are not caught here.

File: Modules/gcmodule.c
Comment (serhiy-storchaka) [inline Modules/gcmodule.c:1070]: Please fix the style of this line too.

File: Modules/gcmodule.c
Comment (1st1) [inline Modules/gcmodule.c:1540]: @serhiy-storchaka Do you understand this `PyGILState_Release` call? Isn't it unsafe to execute Python/CAPI code after it?

File: Modules/gcmodule.c
Comment (serhiy-storchaka) [inline Modules/gcmodule.c:1563]: Maybe it would be better to chain exceptions, but this is an additional non-trivial code. We can do this in beta2.

File: Modules/gcmodule.c
Comment (serhiy-storchaka) [inline Modules/gcmodule.c:1070]: Please fix the style of this line too.

File: Modules/gcmodule.c
Comment (serhiy-storchaka) [inline Modules/gcmodule.c:1544]: You could just return `ret`.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by ned-deily** [comment]
> If there's time, when this one is reviewed by @lisroach and @serhiy-storchaka it would be great if it gets merged into beta-1.

@1st1 Sorry, 3.7.0b1 is finished. Unless there is a consensus that this needs an emergency fix because 3.7.0b1 is unusable, it will have to wait for the scheduled release of 3.7.0b2 in 4 weeks.

File: Modules/gcmodule.c
Comment (gpshead) [reply Modules/gcmodule.c:1540]: here and below, the 

```
    PyGILState_Release(gstate);
    if (ret == NULL) {
        return NULL;
    }
    Py_DECREF(ret);
```

blocks should be replaced with

```
    Py_XDECREF(ret);
    PyGILState_Release(gstate);
    if (ret == NULL) {
        return NULL;
    }
```

The GIL must be held for the decref.

**Review by gpshead — CHANGES_REQUESTED** [review]

File: Modules/gcmodule.c
Comment (1st1) [reply Modules/gcmodule.c:1540]: But this is a Python API!  So we release the gil inside `with ensure_disabled()`. Why?

Comment (serhiy-storchaka) [reply Modules/gcmodule.c:1540]: I didn't review the initial changes and know nothing about the correctness of other code. I'll check it carefully after beta1.

**Comment by 1st1** [comment]
> I don't understand the "test_ensure_disabled_thread" test: why is thread_inside_status is True?

@serhiy-storchaka @gpshead Can you explain this ^^?

File: Modules/gcmodule.c
Comment (1st1) [reply Modules/gcmodule.c:1540]: IMO this is just plain wrong.

**Comment by 1st1** [comment]
> @1st1 Sorry, 3.7.0b1 is finished. Unless there is a consensus that this needs an emergency fix because 3.7.0b1 is unusable, it will have to wait for the scheduled release of 3.7.0b2 in 4 weeks.

@ned-deily Can you run `$ ./python -We -m test test_gc`?  It sig-aborts.  I don't know if that's acceptable or not, your call.

**Comment by 1st1** [comment]
Moreover, currently `with gc.ensure_disabled()` releases GIL for any Python code run inside it.  So it looks like that this is completely broken.

**Comment by ned-deily** [comment]
@1st1 The trap abort is unfortunate but it seems like that it only happens on a debug build which non-developers users will not see.  The question is what impact does this bug have on users.  Can we please move that discussion to bpo-31356 to get a wider audience?

**Comment by 1st1** [comment]
@ned-deily Sure. See new comments from Serhiy and me in the issue.

**Comment by 1st1** [comment]
Closing this one as it was decided that we'll simply revert the original PR instead of fixing it.  Thanks!

**Pull request closed without merging**
